router/src/cluster_based.rs

```diff
@@ -2,23 +2,54 @@
 
 //! A router based on the [`cluster::Cluster`].
 
+use std::time::Duration;
+
 use async_trait::async_trait;
 use ceresdbproto::storage::{Route, RouteRequest};
 use cluster::ClusterRef;
 use common_util::error::BoxError;
 use meta_client::types::RouteTablesRequest;
+use moka::future::Cache;
 use snafu::ResultExt;
 
-use crate::{endpoint::Endpoint, OtherWithCause, ParseEndpoint, Result, Router};
+use crate::{endpoint::Endpoint, OtherWithCause, ParseEndpoint, Result, RouteCacheConfig, Router};
 
 pub struct ClusterBasedRouter {
     cluster: ClusterRef,
+    cache: Cache<String, Route>,
 }
 
 impl ClusterBasedRouter {
-    pub fn new(cluster: ClusterRef) -> Self {
-        Self { cluster }
+    pub fn new(cluster: ClusterRef, cache_config: RouteCacheConfig) -> Self {
+        Self {
+            cluster,
+            cache: Cache::builder()
+                .time_to_live(Duration::from_secs(cache_config.ttl))
+                .time_to_idle(Duration::from_secs(cache_config.tti))
+                .max_capacity(cache_config.capacity)
+                .build(),
+        }
+    }
+}
+
+/// route table from local cache, return cache routes and  tables which are not
+/// in cache
+fn route_from_cache(
+    cache: &Cache<String, Route>,
+    tables: Vec<String>,
+) -> (Vec<Route>, Vec<String>) {
```

Comment:
```suggestion
fn route_from_cache(
    &self,
    tables: Vec<String>,
) -> (Vec<Route>, Vec<String>) {
```

Reply:
> 

This will make the unit test difficult to write.

Reply:
The testing for this function may be limited and it feels like it is testing the functionality of Moka.